File: hedwig-server/src/main/java/com/nicologis/slack/SlackPayload.java
package com.nicologis.slack;
import com.nicologis.teamcity.BuildInfo;
import com.google.gson.annotations.Expose;
import org.apache.commons.lang.StringUtils;

import java.util.ArrayList;
import java.util.List;
import java.util.Map;

public class SlackPayload {
    @Expose
    protected String text;
    @Expose
    protected String channel;
    @Expose
    protected String username;
    @Expose
    protected List<Attachment> attachments;

    private class Attachment {
        @Expose
        protected String fallback;
        @Expose
        protected String pretext;
        @Expose
        protected String color;
        @Expose
        protected List<AttachmentField> fields;
    }

    private class AttachmentField {
        public AttachmentField (String name, String val, boolean isShort) {
            title = name;
            value = val;
            this.isShort = isShort;
        }
        @Expose
        protected String title;
        @Expose
        protected String value;
        @Expose
        protected boolean isShort;
    }

    public void setRecipient(String recipient) {
        this.channel = recipient;
    }

    public void setBotName(String botName) {
        this.username = botName;
    }

    public SlackPayload(BuildInfo build) {
        String branch = build.getBranchName();
        String escapedBranch = branch.length() > 0 ? " [" + branch + "]" : "";
        String statusText = "<" + build.getBuildLink() + "|" + build.getStatusText() + ">";

        String statusEmoji;
        StatusColor statusColor = build.getStatusColor();
        if(statusColor.equals(StatusColor.danger)) {
            statusEmoji = ":x: ";
        }else if(statusColor.equals(StatusColor.warning)){
            statusEmoji = "";
        }else if(statusColor.equals(StatusColor.info)){
            statusEmoji = ":information_source: ";
        }
        else{
            statusEmoji = ":white_check_mark: ";
        }

        String payloadText = statusEmoji + build.getBuildFullName()
                + escapedBranch + " #" + build.getBuildNumber() + " " + statusText;
        this.text = payloadText;

        Attachment attachment = new Attachment();
        attachment.color = statusColor.name();
        attachment.pretext = "Build Information";
        attachment.fallback = payloadText;
        attachment.fields = new ArrayList<>();

        String encodedPrUrl = build.getEncodedPrUrl();
        if(StringUtils.isNotEmpty(encodedPrUrl)){
            AttachmentField prUrlField = new AttachmentField("PullRequest URL", encodedPrUrl, false);
            attachment.fields.add(prUrlField);
        }

        for(Map.Entry<String, String> entry : build.getMessages().entrySet()){
            AttachmentField field = new AttachmentField(entry.getKey(), entry.getValue(), false);
            attachment.fields.add(field);
        }

        this.attachments = new ArrayList<>();
        this.attachments.add(0, attachment);
    }
}

File: hedwig-server/src/main/java/com/nicologis/slack/SlackPayload.kt
package com.nicologis.slack

import com.nicologis.teamcity.BuildInfo
import com.google.gson.annotations.Expose
import org.apache.commons.lang.StringUtils

import java.util.ArrayList

class SlackPayload(build: BuildInfo) {
    @Expose
    private var text: String
    @Expose
    private lateinit var channel: String
    @Expose
    private lateinit var username: String
    @Expose
    private var attachments: MutableList<Attachment>

    inner class Attachment {
        @Expose
        var fallback: String? = null
        @Expose
        var pretext: String? = null
        @Expose
        var color: String? = null
        @Expose
        var fields: MutableList<AttachmentField>? = null
    }

    inner class AttachmentField(
            @field:Expose
            private var title: String,
            @field:Expose
            private var value: String,
            @field:Expose
            private var isShort: Boolean
    )

    fun setRecipient(recipient: String) {
        this.channel = recipient
    }

    fun setBotName(botName: String) {
        this.username = botName
    }

    private fun escape(s: String): String {
        return s.replace("<", "&lt;").replace(">", "&gt;")
    }

    private fun escapeNewline(s: String): String {
        return s.replace("\n", "\\n")
    }

    init {
        val branch = escape(build.branchName)
        val escapedBranch = if (branch.isNotEmpty()) " [$branch]" else ""
        val statusText = ("<" + build.getBuildLink { x -> this.escape(x) }
                + "|" + escape(escapeNewline(build.statusText)) + ">")

        val statusEmoji: String
        val statusColor = build.statusColor
        statusEmoji = when (statusColor) {
            StatusColor.danger -> ":x: "
            StatusColor.warning -> ""
            StatusColor.info -> ":information_source: "
            else -> ":white_check_mark: "
        }

        val payloadText = (statusEmoji + build.buildFullName
                + escapedBranch + " #" + build.buildNumber + " " + statusText)
        this.text = payloadText

        val attachment = Attachment()
        attachment.color = statusColor.name
        attachment.pretext = "Build Information"
        attachment.fallback = payloadText
        attachment.fields = ArrayList()

        val encodedPrUrl = build.getEncodedPrUrl()
        if (StringUtils.isNotEmpty(encodedPrUrl)) {
            val prUrlField = AttachmentField("PullRequest URL", encodedPrUrl, false)
            attachment.fields!!.add(prUrlField)
        }

        for ((key, value) in build.messages) {
            val field = AttachmentField(key, value, false)
            attachment.fields!!.add(field)
        }

        this.attachments = ArrayList()
        if (!attachment.fields!!.isEmpty()) {
            this.attachments.add(0, attachment)
        }
    }
}
